### PR TITLE
[JBWS-4430]:Sever throws IllegalStateException when call a handler wi…

### PR DESCRIPTION
**Pull request by jimma**
…th the CDI bean invocation

### CODE REVIEW

File: modules/server/src/main/java/org/jboss/wsf/stack/cxf/interceptor/HandlerAuthInterceptor.java [hunk 1]
```diff
@@ -55,13 +56,28 @@
 public class HandlerAuthInterceptor extends AbstractPhaseInterceptor<Message>
 {
    private static final String KEY = HandlerAuthInterceptor.class.getName() + ".SECURITY_EXCEPTION";
-   
+   private boolean skipAuth = false;
```
Comment (asoldano) [inline modules/server/src/main/java/org/jboss/wsf/stack/cxf/interceptor/HandlerAuthInterceptor.java:59]: Consider making this `final` if possible:
`private boolean skipAuth;`

and then you set that field in both constructors.

File: modules/server/src/main/java/org/jboss/wsf/stack/cxf/interceptor/HandlerAuthInterceptor.java [hunk 2]
```diff
@@ -91,23 +107,47 @@ private boolean isOutbound(Message message, Exchange ex) {
    private static class JBossWSHandlerChainInvoker extends HandlerChainInvoker
    {
 
-      public JBossWSHandlerChainInvoker(@SuppressWarnings("rawtypes") List<Handler> hc, boolean isOutbound)
+      private boolean skipAuth = false;
```
Comment (asoldano) [inline modules/server/src/main/java/org/jboss/wsf/stack/cxf/interceptor/HandlerAuthInterceptor.java:110]: same as above regarding being `final`

**Review by asoldano — CHANGES_REQUESTED** [review]
hi @jimma , when this will be ready (in general, not just wrt my other comments here), please be sure to include a testcase to cover this new scenario. Thanks!

**Comment by jimma** [comment]
> hi @jimma , when this will be ready (in general, not just wrt my other comments here), please be sure to include a testcase to cover this new scenario. Thanks!

Yes. That is on my todo list.

**Review by asoldano — APPROVED** [review]
Thanks Jim, this looks ok, I've added a couple of minor comments.
WRT the AbstractTCCLPhaseInterceptor, I'm wondering if you explored the possibility of configuring that automatically by interjecting the creation of the interceptor chain in a modified PhaseInterceptorChain which could decide to wrap configured interceptors. Just a thought though, I haven't really verified / tested anything ;-)

File: modules/server/src/main/java/org/jboss/wsf/stack/cxf/interceptor/HandlerAuthInterceptor.java [hunk 3]
```diff
@@ -44,23 +44,41 @@
...
       super(Phase.PRE_PROTOCOL_FRONTEND);
       addBefore(SOAPHandlerInterceptor.class.getName());
       addBefore(LogicalHandlerInInterceptor.class.getName());
+      skip = false;
+   }
+   /**
+    * Create a {@code HandlerAuthInterceptor} that can optionally skip authentication.
+    * When the authentication is skipped, it added a customized {@code JBossWSHandlerChainInvoker}
+    * which set the correct TCCL to allow the handler to access CDI
+    * Please see
```
[the remark this answers is not in the document]
Comment (jimma) [reply modules/server/src/main/java/org/jboss/wsf/stack/cxf/interceptor/HandlerAuthInterceptor.java:72]: Fixed. Thanks.

**Comment by jimma** [comment]
> Thanks Jim, this looks ok, I've added a couple of minor comments. WRT the AbstractTCCLPhaseInterceptor, I'm wondering if you explored the possibility of configuring that automatically by interjecting the creation of the interceptor chain in a modified PhaseInterceptorChain which could decide to wrap configured interceptors. Just a thought though, I haven't really verified / tested anything ;-)

@asoldano  Yes. I tried with checking the classloader and by adding set TCCL with interceptor.getClass().getClassLoader() in PhaseInterceptorChain for each interceptor before the interceptor call,  but this will slows down the performance. The current approach is more efficient as we need to set TCCL only for the interceptor which has CDI access.

**Pull request merged**